### DISCUSSION TITLE
iPXE cannot resolve DNS

### DISCUSSION THREAD

**Discussion by manutenfruits** [ipxe/ipxe#846] [General]
I run netboot.xyz using the docker image, and I keep getting this error (https://ipxe.org/err/3e1161) which seems to be a DNS issue. I have used both a local and a remote (8.8.8.8) DNS and every menu entry has the same issue. Same issue in all hosts (UEFI and legacy boot).

I have tried to debug it with wireshark and I cannot see the host making any kind of DNS requests, which is concerning. Loading over IP seems to work fine.

![image](https://user-images.githubusercontent.com/2961735/209222603-85eac815-908c-4d63-a492-2c043cf5423e.png)

Here's the network config: 
![image](https://user-images.githubusercontent.com/2961735/209222630-efc364b5-926b-4bb2-8507-2dac4b1b2d89.png)

**Comment by NiKiZe** [comment]
One possibility is that your iPXE client can't reach the DNS server at all. How was the client configured in the first place? Does DHCP work? Maybe it has issues with TX or RX, after any of those failures, what does ifstat show?

**Reply by manutenfruits** [reply]
That seems like the most reasonable cause, but I cannot troubleshoot past it. 

All clients that I've used to try (UEFI/Legacy, VM/real machine) get their IP using DHCP. I can ping the DNS host from the iPXE shell. I don't see any issues in the firewall because they're in the same subnet.

I use docker with macvlan so it gets a whole IP, which should rule out some networking issues.

**Reply by NiKiZe** [reply]
Then firewall on some or all UDP traffic, please read the error page carefully ;)

**Reply by stappersg** [reply]
I did read that _Original Poster_ did not see any DNS traffic with `wireshark`. So now I wonder how the _wiresharking_ was done. If the outcome is something like _"closer the iPXE client does `wireshark` see DNS traffic, a firewall might causing this problem"_ is fine. If the outcome is _"fairly sure that iPXE doesn't sent out DNS request"_ we can zoom in on that.  (And main stakeholder of that _we_ is _O.P._.)

**Reply by manutenfruits** [reply]
> Then firewall on some or all UDP traffic, please read the error page carefully ;)

Can you be more specific? I have rules set to let all traffic on that lan pass through.

> So now I wonder how the _wiresharking_ was done.

Fair point. I used `wireshark -k -i <(ssh root@10.20.0.1 tcpdump -i vtnet1 -U -w - not tcp port 22)` (I use OPNSense as the router/fw. vtnet1 is the interface for the lan. I do see the DHCP traffic when using `dhcp` as filter, but I don't see much else when I use `ip.src==<pxe_client_ip>`, ~~even though the menu loads fine (so there is definitely some traffic that isn't being picked up). Maybe the wiresharking is wrong?~~

Right, I cannot see traffic between clients in the lan because that traffic doesn't hit the firewall unless it's going to another subnet. However, it should see _something_ when making DNS requests as those can hit the firewall itself (local DNS), or public DNS.

I'll look into capturing packets more accurately, but my point about not seeing DNS traffic stands.

**Reply by manutenfruits** [reply]
Okay! I was able to capture traffic with a Proxmox VM on its tap interface. Looks like the DNS request is done over IPv6 (which would explain no traffic when filtering by IP), and it does get rejected. I don't know why it's using IPv6 though, because I have it disabled on the router interface, no DHCP is enabled for v6, and the network page of netboot.xyz shows IPv4 addresses, as expected.

![image](https://user-images.githubusercontent.com/2961735/209337508-132519c1-9524-402d-b9cf-071ae987d8af.png)

**Reply by mcb30** [reply]
@manutenfruits You have an IPv6 router issuing IPv6 Unique Local Addresses within the subnet.  The address used is visibly not a SLAAC address, which means that the router is instructing iPXE to use DHCPv6, and you have a DHCPv6 server providing stateful DHCPv6 address assignment.

Either the DHCPv6 server or the router (or both) is handing out the address `fd76:744:802::1` as a DNS server.  iPXE is respecting this setting and will attempt to query this DNS server first (since IPv6 DNS servers are given precedence to IPv4 DNS servers).

The DNS server `fd76:744:802::1` is refusing the DNS query, presumably due to a misconfiguration in your setup.  This should cause iPXE to fail over to the next DNS server in the list (which will probably be the IPv4 DNS server), but this does not seem to be happening.

**Reply by stappersg** [reply]
![afbeelding](https://user-images.githubusercontent.com/35434094/209339815-31f0d635-f569-4a33-b68e-5262d311a218.png)

For me is this discussion about _"iPXE cannot resolve DNS"_ completed.

**Reply by manutenfruits** [reply]
Thank you! My router had it disabled but there was a second openwrt that I use as an AP that was going rogue and responding to DHCPv6 requests and wreaking havoc. Now it all works 😍

**Reply by romanchyla** [reply]
thank you @manutenfruits I've hit the same problem and your post has helped me figure it out. I also had ipv6 disabled on my router, and two "dumb AP" running openwrt. One of those closer to the switch (shorter cable length) was confusing the nodes. I don't know the real cause yet, but the symptoms were the same: ipxe would boot, it had connectivity (ping 1.1.1.1 worked), but DNS queries didn't resolve. When I tried dhcpdump - on another machine -- no  ipv6 stuff was being sent out, but it looks like ipv4 and ipv6 dhcp responses can arrive from separate dhcp servers...

It is curious that iPXE doesn't try ivp4 dns resolution when ipv6 fails. Seems a problem known to have been there for 5 years: https://ipxe-devel.ipxe.narkive.com/TmVy2Mdd/fallback-to-ipv4-or-disable-ipv6

**Comment by stappersg** [comment]
The

>   ... which seems to be a DNS issue ...

and 

>  ... cannot see the host making any kind of DNS requests, 

do match.


> which is concerning.

I dare to disagree  :smiley:




> Loading over IP seems to work fine.

Please elaborate.

**Comment by manutenfruits** [comment]
I mean that if I use IP addresses instead of URLs with hostnames to fetch the booting files, then it works. So that narrows it down to a DNS resolution issue.

I said it was concerning because if it's not even making DNS requests because I'm not seeing them in wireshark, then I have little faith in trying different config values.

**Reply by stappersg** [reply]
Acknowledge.

The _iPXE_ that is being used, does it have DNS code included?  Share with us the full version banner of _iPXE_.

**Comment by jamesaepp** [comment]
Wanted to comment here that this was still useful/helpful as of today. v6 stack taking precedence over v4 stack explained away my issue with the same error code trying to chain an HTTP file by DNS while chaining the file via IPv4 worked fine.

**Comment by minilandl** [comment]
Just ran into this issue as well today this was very helpful as I spent several hours trying to work it out. I also had an openwrt ap handing out rogue ipv6 addresses disabled ipv6 and I was able to load images via netboot.xyz

**Comment by illsk1lls** [comment]
I had a similar issue, intermittently I do not receive any DNS query from iPXE(the client) after DHCP, i send the server IP (DNS) as option 6

My setup HTTP/DHCP/DNS/TFTP are all in a single powershell script, i can see all the traffic in wireshark so I know it's not on my end, but sometimes a DNS request just never comes.

`ipxe/src/config/general.h` i have set `#undef NET_PROTO_IPV6` as well as in `ipxe/src/config/efi.h` and `ipxe/src/config/pcbios.h`

With that in mind I don't think it is a IPv6 issue unless I have missed something.

For my setup currently i use embedded `boot.ipxe` and `autoexec.ipxe` in `undionly.kpxe` and `ipxe.efi` respectively

they both retry dhcp if dns isnt working, e.g.

```
#!ipxe

:startup
dhcp
kernel http://pxeserver.local/wimboot || goto startup
initrd http://pxeserver.local/BCD BCD
initrd http://pxeserver.local/boot.sdi boot.sdi
initrd http://pxeserver.local/bootmgr.exe bootmgr.exe
initrd http://pxeserver.local/boot.wim boot.wim
boot
```

on occasion it will fail then be fine after the first retry, without adding the retry into the script it would simply fail

the only reason I'm mentioning this is I am making a "portable" pxeserver and i dont know what kind of broadcasts users will be encountering in their environments, i have to be able to switch PXEServer IP's even though I am using a subnet that shouldnt interfere with anything, 169.168.2.x, so DNS is absolutely necessary, and if there is any way to bake in some type of manual DNS request via IPv4 during the boot script that would be awesome, I can work with the source a little but obviously you guys are the experts

**Reply by NiKiZe** [reply]
If you have issues with IPv6, fix the issues, don't disable, don't assume that anyone is even using legacy IPv4 or having a IPv4 enabled network.

There is a set of ipranges available for internal networks, just don't take a random IPv4 subnet. (https://db-ip.com/all/169.168.2)

If iPXE does fail to do a DNS lookup, you will get an error message, go from there, not from "the server isn't seeing the lookup"

And yes, you can configure DNS and all other parameters by embedded scripts.

And finally, as already mentioned, your wimboot script is based on legacy wimboot from a time it wasn't able to extract the needed files itself. Make sure your .wim does contain the required files instead.

**Reply by illsk1lls** [reply]
The wim does not contain the files im using initrd for

I'm currently changing dns.c to fallback onto the DHCP server address for name resolution if no DNS info is received

EDIT: I found my issue it was a competing server, the DEBUG=dhcp,dns during compilation made it visible

**Reply by NiKiZe** [reply]
You can only have one DHCP server on a network, anything else is out of spec.

**Reply by illsk1lls** [reply]
The server listens only for/responds only to PXEClient

And the client now listens for/responds only to PXEServer

bypassed the spec ;) you could have 5 ghost dhcp servers running and it wouldnt matter, ip/dns always gets assigned correctly

it's just an idea, it works really well i've been testing it on my end without fail, and i have been trying to break it

**Reply by NiKiZe** [reply]
Again, it violates RFC don't do this.

**Comment by illsk1lls** [comment]
@NiKiZe, I am working with the older version [e7595fe](https://github.com/ipxe/ipxe/commit/e7595fe88d8c32354cd61055cbc9f2e7b3f91ff8) due to the CPIO magic issues on some legacy boot systems [#1422](https://github.com/ipxe/ipxe/issues/1422), so I'm not sure how much dhcp has changed in the latest..

But for the DNS issue, and other routers/devices causing connection issues, why dont you allow for a DHCP option code to contain a string to identify the server? That way the client could always find the right server?

In my case I used the string PXEServer w/ 200
`src/net/udp/dhcp.c`

```
static void dhcp_discovery_rx ( struct dhcp_session *dhcp,
                struct dhcp_packet *dhcppkt,
                struct sockaddr_in *peer, uint8_t msgtype,
                struct in_addr server_id,
                struct in_addr pseudo_id ) {
    struct in_addr ip;
    char vci[9]; /* "PXEClient" */
    int vci_len;
    int has_pxeclient;
    int8_t priority = 0;
    uint8_t no_pxedhcp = 0;
    unsigned long elapsed;
    uint8_t option_data[16]; 
    size_t len;

    DBGC ( dhcp, "DHCP %p %s from %s:%d", dhcp,
           dhcp_msgtype_name ( msgtype ), inet_ntoa ( peer->sin_addr ),
           ntohs ( peer->sin_port ) );
    if ( ( server_id.s_addr != peer->sin_addr.s_addr ) ||
         ( pseudo_id.s_addr != peer->sin_addr.s_addr ) ) {
        DBGC ( dhcp, " (%s/", inet_ntoa ( server_id ) );
        DBGC ( dhcp, "%s)", inet_ntoa ( pseudo_id ) );
    }

    /* Check for option 200 with "PXEServer" */
    len = dhcppkt_fetch ( dhcppkt, 200, option_data, sizeof ( option_data ) );
    if ( len != strlen("PXEServer") || memcmp(option_data, "PXEServer", len) != 0 ) {
        DBGC ( dhcp, "DHCP %p rejecting packet: missing or invalid option 200 (expected 'PXEServer')\n", dhcp );
        return; /* Discard the packet */
    }

    /* Identify offered IP address */
    ip = dhcppkt->dhcphdr->yiaddr;
    if ( ip.s_addr )
        DBGC ( dhcp, " for %s", inet_ntoa ( ip ) );

    /* Identify "PXEClient" vendor class */
    vci_len = dhcppkt_fetch ( dhcppkt, DHCP_VENDOR_CLASS_ID,
                  vci, sizeof ( vci ) );
    has_pxeclient = ( ( vci_len >= ( int ) sizeof ( vci ) ) &&
              ( strncmp ( "PXEClient", vci, sizeof (vci) ) == 0 ));
    if ( has_pxeclient ) {
        DBGC ( dhcp, "%s",
               ( dhcp_has_pxeopts ( dhcppkt ) ? " pxe" : " proxy" ) );
    }

    /* Identify priority */
    dhcppkt_fetch ( dhcppkt, DHCP_EB_PRIORITY, &priority,
            sizeof ( priority ) );
    if ( priority )
        DBGC ( dhcp, " pri %d", priority );

    /* Identify ignore-PXE flag */
    dhcppkt_fetch ( dhcppkt, DHCP_EB_NO_PXEDHCP, &no_pxedhcp,
            sizeof ( no_pxedhcp ) );
    if ( no_pxedhcp )
        DBGC ( dhcp, " nopxe" );
    DBGC ( dhcp, "\n" );

    /* Select as DHCP offer, if applicable */
    if ( ip.s_addr && ( peer->sin_port == htons ( BOOTPS_PORT ) ) &&
         ( ( msgtype == DHCPOFFER ) || ( !msgtype /* BOOTP */ ) ) &&
         ( priority >= dhcp->priority ) ) {
        dhcp->offer = ip;
        dhcp->server = server_id;
        dhcp->priority = priority;
        dhcp->no_pxedhcp = no_pxedhcp;
    }

    /* Select as ProxyDHCP offer, if applicable */
    if ( pseudo_id.s_addr && has_pxeclient &&
         ( priority >= dhcp->proxy_priority ) ) {
        dhcppkt_put ( dhcp->proxy_offer );
        dhcp->proxy_server = pseudo_id;
        dhcp->proxy_offer = dhcppkt_get ( dhcppkt );
        dhcp->proxy_priority = priority;
    }

    /* We can exit the discovery state when we have a valid
     * DHCPOFFER, and either:
     *
     *  o  The DHCPOFFER instructs us to ignore ProxyDHCPOFFERs, or
     *  o  We have a valid ProxyDHCPOFFER, or
     *  o  We have allowed sufficient time for ProxyDHCPOFFERs.
     */

    /* If we don't have a DHCPOFFER, do nothing */
    if ( ! dhcp->offer.s_addr )
        return;

    /* If we can't yet transition to DHCPREQUEST, do nothing */
    elapsed = ( currticks() - dhcp->start );
    if ( ! ( dhcp->no_pxedhcp || dhcp->proxy_offer ||
         ( elapsed > DHCP_DISC_PROXY_TIMEOUT_SEC * TICKS_PER_SEC ) ) )
        return;

    /* Transition to DHCPREQUEST */
    dhcp_set_state ( dhcp, &dhcp_state_request );
}
```

This allows me to get through the noise by ignoring everything that isn't identifying itself as my server.
